File: 2021/fsharp/Aoc2021/Day10.fs
module Day10

open Utils

let isOpening =
    function
    | '('
    | '['
    | '{'
    | '<' -> true
    | _ -> false

let other =
    function
    | '(' -> ')'
    | '[' -> ']'
    | '{' -> '}'
    | '<' -> '>'
    | ')' -> '('
    | ']' -> '['
    | '}' -> '{'
    | '>' -> '<'
    | _ -> 0 |> char

let syntaxErrorScore line =
    line
    |> Seq.fold
        (fun (score, openings) char ->
            match score with
            | 0 ->
                if char |> isOpening then
                    score, char :: openings
                else
                    let expected = openings |> List.head |> other

                    let score =
                        if char = expected then
                            score
                        else
                            match char with
                            | ')' -> 3
                            | ']' -> 57
                            | '}' -> 1197
                            | '>' -> 25137
                            | _ -> -1

                    score, openings |> List.tail
            | _ -> score, openings)
        (0, [])
    |> fst

let completionScore line =
    line
    |> Seq.fold
        (fun openings char ->
            if char |> isOpening then
                char :: openings
            else
                openings |> List.tail)
        List.empty
    |> Seq.map other
    |> Seq.fold
        (fun score char ->
            let points =
                match char with
                | ')' -> 1L
                | ']' -> 2L
                | '}' -> 3L
                | '>' -> 4L
                | _ -> 0L

            score * 5L + points)
        0L

let totalSyntaxErrorScore lines =
    lines |> List.map syntaxErrorScore |> List.sum

let middleCompletion lines =
    let scores =
        lines
        |> List.filter (syntaxErrorScore >> (fun score -> score = 0))
        |> List.map completionScore
        |> List.sort

    scores |> List.item (List.length scores / 2)

let input = "10" |> readInputLines |> Seq.toList

let job1 () =
    input |> totalSyntaxErrorScore |> string

let job2 () = input |> middleCompletion |> string

File: 2021/csharp/Aoc2021/Day10.cs
namespace Aoc2021;

public class Day10 : IDay
{
    public override string Day { get; } = nameof(Day10)[3..];

    public override string Result1()
    {
        return TotalSyntaxErrorScore(InputLines)
            .ToString(CultureInfo.InvariantCulture);
    }

    public override string Result2()
    {
        return MiddleCompletionScore(InputLines)
            .ToString(CultureInfo.InvariantCulture);
    }

    public static int TotalSyntaxErrorScore(IEnumerable<string> lines)
    {
        return lines
            .Select(SyntaxErrorScore)
            .Sum();
    }

    public static long MiddleCompletionScore(IEnumerable<string> lines)
    {
        List<long> scores = lines
            .Where(line => SyntaxErrorScore(line) == 0)
            .Select(CompletionScore)
            .ToList();
        scores.Sort();
        return scores[scores.Count / 2];

        static long CompletionScore(string line)
        {
            return line
                .Aggregate(new List<char>(), (openings, c) =>
                {
                    if (c is '(' or '[' or '{' or '<')
                    {
                        openings.Add(c);
                    }
                    else
                    {
                        openings.RemoveAt(openings.Count - 1);
                    }
                    return openings;
                })
                .Select(c => c switch
                {
                    '(' => ')',
                    '[' => ']',
                    '{' => '}',
                    '<' => '>',
                    _ => '\0'
                })
                .Reverse()
                .Aggregate(0L, (score, c) =>
                {
                    long points = c switch
                    {
                        ')' => 1L,
                        ']' => 2L,
                        '}' => 3L,
                        '>' => 4L,
                        _ => '\0'
                    };
                    return (score * 5L) + points;
                });
        }
    }

    private static int SyntaxErrorScore(string line)
    {
        List<char> openings = new();
        foreach (char c in line)
        {
            if (c is '(' or '[' or '{' or '<')
            {
                openings.Add(c);
                continue;
            }

            char expectedClosing = openings.Last() switch
            {
                '(' => ')',
                '[' => ']',
                '{' => '}',
                '<' => '>',
                _ => '\0'
            };
            if (c != expectedClosing)
            {
                return c switch
                {
                    ')' => 3,
                    ']' => 57,
                    '}' => 1197,
                    '>' => 25137,
                    _ => throw new ApplicationException("Invalid char"),
                };
            }
            openings.RemoveAt(openings.Count - 1);
        }
        return 0;
    }
}
